bottomless/src/wal.rs

```diff
@@ -234,6 +253,47 @@ pub fn checksum_be(init: u64, page: &[u8]) -> u64 {
     ((s1 as u64) << 32) | (s2 as u64)
 }
 
+// 128KiB since many OSes already prefetch 128KiB and that's also unit of pricing in S3
+const COPY_BUF_SIZE: u64 = 128 * 1024;
+
+/// Copies up to `len` bytes from `src` to `dst`. Returns number of written bytes,
+/// which can differ from bytes read eg. when compression is used.
+async fn copy_range<R, W>(src: &mut R, dst: &mut W, len: u64) -> Result<u64>
```

Comment:
is this needed because `tokio::io::copy` always reads the whole reader, but in this case we want to specifically stop after `len` bytes?

In that case, perhaps let's still use `tokio::io::copy`, but with the reader wrapped in https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.take?